File: back/HistoryBank/src/main/java/com/majortomdev/historybank/service/UserService.java
package com.majortomdev.historybank.service;

import com.majortomdev.historybank.dto.LoginDto;
import com.majortomdev.historybank.dto.UserDto;
import com.majortomdev.historybank.entities.User;
import com.majortomdev.historybank.repository.UserRepository;
import com.majortomdev.historybank.util.JwtUtil;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtUtil jwtUtil;

    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder, JwtUtil jwtUtil) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtUtil = jwtUtil;
    }

    public User register(UserDto userDto) {
        if (userRepository.findByUsername(userDto.getUsername()) != null) {
            throw new RuntimeException("User already exists");
        }

        String encodedPassword = passwordEncoder.encode(userDto.getPassword());
        User user = new User(userDto.getUsername(), encodedPassword);
        return userRepository.save(user);
    }

    public String login(LoginDto loginDto) {
        User user = userRepository.findByUsername(loginDto.getUsername());
        if (user == null) {
            throw new RuntimeException("User not found");
        }

        if (!passwordEncoder.matches(loginDto.getPassword(), user.getPassword())) {
            throw new RuntimeException("Invalid credentials");
        }

        return jwtUtil.generateToken(user.getUsername()); // Generate JWT token
    }

//    public User findUserByUsername(String username) {
//        return userRepository.findByUsername(username)
//                .orElseThrow(() -> new RuntimeException("User not found"));
//    }
}

File: back/HistoryBank/src/main/kotlin/com/majortomdev/HistoryBank/Service/UserService.kt
@Service
class UserService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtUtil: JwtUtil
) {

    fun register(userDto: UserDto): User {
        if (userRepository.findByUsername(userDto.username) != null) {
            throw RuntimeException("User already exists")
        }

        val encodedPassword = passwordEncoder.encode(userDto.password)
        val user = User(username = userDto.username, password = encodedPassword)
        return userRepository.save(user)
    }

    fun login(loginDto: LoginDto): String {
        val user = userRepository.findByUsername(loginDto.username)
            ?: throw RuntimeException("User not found")

        if (!passwordEncoder.matches(loginDto.password, user.password)) {
            throw RuntimeException("Invalid credentials")
        }

        return jwtUtil.generateToken(user.username)// Generate JWT token
    }

    fun findUserByUsername(username: String): User {
        return userRepository.findByUsername(username)
            ?: throw RuntimeException("User not found")
    }
}
